### PR TITLE
feat(profiles): cache historic exchange rates

### PR DESCRIPTION
**Pull request by goga-m**
#### Caching  
- [x] Keeps daily cached rate pairs for a defined duration N in milliseconds. 
- [x] All rates are stored in ExchangeRate dataRepository
- [x] Custom cached duration can be added in constructor: `new ExchangeRateService({ ttl: N})`

#### Storage
- [x] Refactored `ExchangeRateService#syncCoinByProfile()` to store daily rates for all coins in a single `ExchangeRateService#dataStorage` instead of storing them in each wallet
- [x] Added `ExchangeRateService#allRates()`
- [x] Added `ExchangeRateService#ratesByDate(currency: string, exchangeCurrency: string, date?: string | number | Datetime)`
#### Wallet
- [x] Feed `Wallet#exchangeRate()` from ExchangeRateService, drop storing/pulling rates from `wallet.data().get(Identity.ExchangeRate)`
- [x] Use exchange rates in `Wallet#convertedBalance()`
- [x] `Wallet#exchangeCurreny()` to pull exchange currency from profile instead of `wallet.data()` 

## Checklist

- [ ] Documentation _(if necessary)_
- [x] Tests _(if necessary)_
- [x] Ready to be merged

### CODE REVIEW

File: packages/platform-sdk-profiles/src/environment/services/exchange-rate-service.ts [hunk 1]
```diff
@@ -7,8 +7,18 @@ import { ProfileRepository } from "../../repositories/profile-repository";
 import { ReadWriteWallet, WalletData } from "../../wallets/wallet.models";
 import { container } from "../container";
 import { Identifiers } from "../container.models";
+import { DateTime } from "@arkecosystem/platform-sdk-intl";
 
 export class ExchangeRateService {
+	readonly #lastSynced: Record<string, DateTime> = {};
+	readonly #cacheDurationMilliseconds: number = 10000;
+
+	public constructor(options?: { cacheDurationMilliseconds?: number }) {
```
Comment (faustbrian) [inline packages/platform-sdk-profiles/src/environment/services/exchange-rate-service.ts:16]:
```suggestion
	public constructor(options?: { ttl?: number }) {
```

File: packages/platform-sdk-profiles/src/environment/services/exchange-rate-service.ts [hunk 2]
```diff
@@ -7,8 +7,18 @@ import { ProfileRepository } from "../../repositories/profile-repository";
 import { ReadWriteWallet, WalletData } from "../../wallets/wallet.models";
 import { container } from "../container";
 import { Identifiers } from "../container.models";
+import { DateTime } from "@arkecosystem/platform-sdk-intl";
 
 export class ExchangeRateService {
+	readonly #lastSynced: Record<string, DateTime> = {};
+	readonly #cacheDurationMilliseconds: number = 10000;
```
Comment (faustbrian) [inline packages/platform-sdk-profiles/src/environment/services/exchange-rate-service.ts:14]: Treat the ttl in seconds. We are never caching it based on milliseconds so we should stick to seconds or minutes for the ttl.

File: packages/platform-sdk-profiles/src/environment/services/exchange-rate-service.ts [hunk 3]
```diff
@@ -36,17 +46,43 @@ export class ExchangeRateService {
 			return;
 		}
 
+		const exchangeCurrency: string = profile.settings().get(ProfileSetting.ExchangeCurrency) || "BTC";
+		if (!this.shouldSync(exchangeCurrency)) return;
```
Comment (faustbrian) [inline packages/platform-sdk-profiles/src/environment/services/exchange-rate-service.ts:50]: Always use bracket syntax.

File: packages/platform-sdk-profiles/src/wallets/wallet.models.ts [hunk 4]
```diff
@@ -42,6 +42,7 @@ export enum WalletData {
 	Delegates = "DELEGATES",
 	ExchangeCurrency = "EXCHANGE_CURRENCY",
 	ExchangeRate = "EXCHANGE_RATE",
+	ExchangeRates = "HISTORICAL_EXCHANGE_RATE",
```
Comment (faustbrian) [inline packages/platform-sdk-profiles/src/wallets/wallet.models.ts:45]:
```suggestion
	ExchangeRates = "EXCHANGE_RATES",
```

The value should always be the `upper-case snake_case` of the key.

File: packages/platform-sdk-profiles/src/environment/services/exchange-rate-service.ts [hunk 5]
```diff
@@ -36,17 +46,43 @@ export class ExchangeRateService {
...
-		const exchangeCurrency: string = profile.settings().get(ProfileSetting.ExchangeCurrency) || "BTC";
 		const exchangeRate = await marketService.dailyAverage(currency, exchangeCurrency, +Date.now());
 
+		this.#lastSynced[exchangeCurrency] = DateTime.make();
+		const date = DateTime.make().format("YYYY-MM-DD");
+
 		for (const wallet of wallets) {
-			wallet.data().set(WalletData.ExchangeCurrency, exchangeCurrency);
-			wallet.data().set(WalletData.ExchangeRate, exchangeRate);
+			this.updateWalletExchangeData(wallet, exchangeCurrency, exchangeRate, date);
```
Comment (faustbrian) [inline packages/platform-sdk-profiles/src/environment/services/exchange-rate-service.ts:62]: This is a promise but not awaited. You should `Promise.all` or `Promise.allSettled` here to get this done fast without having unresolved promises.

Comment (goga-m) [reply packages/platform-sdk-profiles/src/environment/services/exchange-rate-service.ts:64]: @faustbrian  was incorrectly set as promise. `data().get` and `data.set` are sync operations. Fixed it

**Comment by faustbrian** [comment]
Will clean this up and finish it in a separate PR.

**Pull request merged**